SVD-based preconditioners for tall matrices

rileyjmurray opened the pull request:
This PR introduces functionality for computing approximate orthogonalizers of tall data matrices. The implementation is purely procedural, operates on raw buffers (no std::vector), and only allows SASOs for sketching operators. Basic documentation and tests are provided.

This PR also brings RandLAPACK's CI and the RandBLAS submodule up to date to match RandBLAS' main.


#### Note
This PR originally intended to define a helper function for Nystrom preconditioning. That is no longer in this PR's scope. I noticed that our current implementation of low-rank approximation via Nystrom assumes that the symmetric matrix is stored in a full format rather than just upper or lower triangular. I'll make a separate PR that fixes this issue and others first thing this upcoming week.

PTNobel commented:
I'm planning to redesign the preconditioners code in my codebase this week based on the stuff Burlen and I were chatting about today; don't worry about that code for now...

PTNobel commented on RandLAPACK/comps/rl_preconditioners.hh:
```diff
@@ -0,0 +1,208 @@
+        lda = m;
+        lda_sk = d;
+    }
+    blas::scal(d*n, 0.0, buff_A_sk, 1);
```
Question, if I initialize  the buffer to/get unlucky and allocate a page of all NaN before execution, what happens?

Unlike other operations in blaspp, the docs for blas::scal do not disavow that the impl will read the data when alpha = 0

PTNobel replied:
Thinking about this more; in single precision arithemetic there are 2**23 NaNs, so if we drew the initial value of the buffer from a uniform distribution over bit patterns there is in expectation d * n/512 NaNs. Which since NaN * 0 = NaN, means that this buffer for any reasonable sized matrix probably includes at least one NaN if the source of the initial condition is random. If we're dependent on this buffer being 0, I don't think this works.

rileyjmurray replied:
Fixed! Good catch.

PTNobel commented on RandLAPACK/comps/rl_preconditioners.hh:
```diff
@@ -0,0 +1,208 @@
+    );
+
+    // step 2: compute SVD of sketch
+    T* ignore;
```
Explicitely initializing this to NULL might be nice so when debugging code seeing the NULL pointers below looks intended where as we might just get a pointer to a strange part of memory that has nonsensical data otherwise

rileyjmurray replied:
Fixed.

PTNobel commented on RandLAPACK/comps/rl_preconditioners.hh:
```diff
@@ -0,0 +1,208 @@
+    } else {
+        lapack::Job jobu = lapack::Job::NoVec;
+        lapack::Job jobvt = lapack::Job::SomeVec;
+        T *Vt = new T[n * n];
```
Hmmm I don't follow what the allocation is needed for if we specified that the output of this function is in the same layout as `A` could we get away with not allocating?

Oh never mind we need the scratchspace too for this call...

Still feel like some discussion of a "large" allocation might be worthwhile

rileyjmurray replied:
Agreed. I'll do that before we merge.

rileyjmurray replied:
Fixed.

PTNobel commented on RandLAPACK/comps/rl_preconditioners.hh:
```diff
@@ -0,0 +1,208 @@
+}
+
+template <typename T>
+int64_t rpc_svd_saso(
```
No docs?

rileyjmurray replied:
Not yet, lol. Will add before merging.

rileyjmurray replied:
This function has been removed. Accessing equivalent functionality now requires two function calls. The first call is to ``rpc_data_svd_saso``, and the second call is to ``make_rpc_svd_explicit``. The latter of these functions is deterministic, its name is subject to change, and documentation is forthcoming.

PTNobel commented on RandLAPACK/comps/rl_preconditioners.hh:
```diff
@@ -0,0 +1,208 @@
+    int64_t n, // number of columns in A
+    int64_t d, // number of rows in sketch of A
+    int64_t k, // number of nonzeros in each column of the sketching operator
+    std::vector<T>& A, // buffer of size m*n.
```
I can't use an std::vector API without a lot of copies. Is it possible to have a `T *` or `hamr::buffer<T>` API?

rileyjmurray replied:
Sure thing. This function doesn't benefit from memory tricks with std::vector anyway.

PTNobel commented on RandLAPACK/comps/rl_preconditioners.hh:
```diff
@@ -0,0 +1,208 @@
+ *      "layout_A" order.
+ * @param[out] V_sk
+ *      We require that V_sk.size() >= d*n. Letting "r" denote the return
+ *      value of this function, on exit the orthogonal factor of the preconditioner
```
This function returns void. Unclear what r represents.

rileyjmurray replied:
Ah crap. Will fix.

rileyjmurray replied:
Fixed. I've updated the doc to not mention this rank parameter and to have this function return an RNGState.

rileyjmurray commented on RandLAPACK/comps/rl_preconditioners.hh:
```diff
@@ -0,0 +1,208 @@
+ *      It makes M'GM well-conditioned, so that M'GM z = M'h
+ *      can easily be solved by CG without preconditioning.
+ * 
+ * The input matrix A must be contiguous; no "LDA" is allowed.
```
Fixed!

rileyjmurray commented on RandLAPACK/comps/rl_preconditioners.hh:
```diff
@@ -0,0 +1,208 @@
+
+/**
+ * Use sketching to produce data for a right-preconditioner for tall
+ * matrices of the form A_aug = [A; sqrt(mu) * I]. This is useful for ...
```
Fixed

rileyjmurray commented:
@PTNobel, I believe I've addressed all of your comments. When you get a chance, please re-review and request changes or approve as appropriate.

Thanks!

PTNobel commented:
Do you still want a rereview?

rileyjmurray commented:
Oh shoot. I thought the thumbs up was an approval. Please feel free to
review. We can open bug reports for any nontrivial things you identify.

On Mon, May 22, 2023 at 3:06 PM Parth Nobel ***@***.***>
wrote:

> Do you still want a rereview?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/BallisticLA/RandLAPACK/pull/38#issuecomment-1558090719>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACRLIFHPH5TWAAVIM63HDU3XHPPNNANCNFSM6AAAAAAYICZDUM>
> .
> You are receiving this because you modified the open/close state.Message
> ID: ***@***.***>
>
-- 
Riley

PTNobel commented on RandLAPACK/comps/rl_preconditioners.hh:
```diff
@@ -0,0 +1,214 @@
+ * 
+ *      On exit, the i-th column of V is scaled down by
+ *              1/sqrt(sigma[i]^2 + mu)
+ *      up until i = rank-1, where rank is the return value of this function.
```
Indexing here is a little confusing... I assume this means i = 0, ... , rank - 1 but i do worry that the mix of 0- and 1- indexing in math libraries means we may want to just do a small phrasing change to 
```
 *      On exit, the i-th column of V is scaled down by
 *              1/sqrt(sigma[i]^2 + mu)
 *      for i = 0, 1, ..., rank-1, where rank is the return value of this function.
```

rileyjmurray replied:
Excellent observation. I'll make this change.

PTNobel commented on RandLAPACK/comps/rl_preconditioners.hh:
```diff
@@ -0,0 +1,214 @@
+
+/**
+ * Accepts the right singular vectors and singular values of some tall
+ * matrix "T", along with a regularization parameter mu.
```
I'm not a huge fan of the matrix T since T is a type variable for this function and is not any of the parameters, so it might be a bit confusing

rileyjmurray replied:
Good catch!

PTNobel reviewed (commented):
Overall looks great. I have two comments on the docs.

Sorry about the ambiguity for the thumbs up!
